src/mlpack/methods/reinforcement_learning/q_learning.hpp

```diff
@@ -0,0 +1,170 @@
+ *               Alex Graves and
+ *               Ioannis Antonoglou and
+ *               Daan Wierstra and
+ *               Martin A. Riedmiller},
+ *  title     = {Playing Atari with Deep Reinforcement Learning},
+ *  journal   = {CoRR},
+ *  year      = {2013},
+ *  url       = {http://arxiv.org/abs/1312.5602}
+ * }
+ * @endcode
+ *
+ * @tparam EnvironmentType The environment of the reinforcement learning task.
+ * @tparam NetworkType The network to compute action value.
+ * @tparam OptimizerType The optimizer to train the network.
+ * @tparam PolicyType Behavior policy of the agent.
+ * @tparam ReplayType Experience replay method.
+ */
+template <
+  typename EnvironmentType,
+  typename NetworkType,
+  typename OptimizerType,
+  typename PolicyType,
+  typename ReplayType = RandomReplay<EnvironmentType>
+>
+class QLearning
+{
+ public:
+  //! Convenient typedef for state.
+  using StateType = typename EnvironmentType::State;
+
+  //! Convenient typedef for action.
+  using ActionType = typename EnvironmentType::Action;
+
+  /**
+   * Create the QLearning object with given settings.
+   *
+   * @param network The network to compute action value.
+   * @param optimizer The optimizer to train the network.
+   * @param discount Discount for future return.
+   * @param policy Behavior policy of the agent.
+   * @param replayMethod Experience replay method.
+   * @param targetNetworkSyncInterval Interval (steps) to sync the target network.
+   * @param explorationSteps Steps before starting to learn.
+   * @param doubleQLearning Whether to use double Q-Learning.
+   * @param stepLimit Maximum steps in each episode, 0 means no limit.
+   * @param environment Reinforcement learning task.
+   */
+  QLearning(NetworkType& network,
+            OptimizerType& optimizer,
+            double discount,
+            PolicyType policy,
+            ReplayType replayMethod,
+            size_t targetNetworkSyncInterval,
+            size_t explorationSteps,
+            bool doubleQLearning = false,
+            size_t stepLimit = 0,
+            EnvironmentType environment = EnvironmentType());
+
+  /**
+   * Execute a step in an episode.
+   * @return Reward for the step.
+   */
+  double Step();
+
+  /**
+   * Execute an episode.
+   * @return Return of the episode.
+   */
+  double Episode();
+
+  /**
+   * @return Total steps from beginning.
+   */
+  const size_t& TotalSteps() const { return totalSteps; }
+
+  //! Modify the training mode / test mode indicator.
+  bool& Deterministic() { return deterministic; }
+
+  //! Get the indicator of training mode / test mode.
+  const bool& Deterministic() const { return deterministic; }
+
+ private:
+  /**
+   * Select the best action based on given action value.
+   * @param actionValues Action values.
+   * @return Selected actions.
+   */
+  arma::icolvec BestAction(const arma::mat& actionValues);
+
+  //! Reference of the learning network.
+  NetworkType& learningNetwork;
```

Comment:
Getting the model from the optimizer looks very anti-intuitive for me and doesn't solve the problem from root. What I think is the most elegant way is to design a mechanism to allow rebinding for optimizer or delayed binding. It's even more necessary for async methods -- each thread will have its own model and may have their own optimizer, so the model must be bound to its corresponding optimizer dynamically. Currently it seems we cannot do this.
I have some solutions, one is delayed binding by lambda function. In stead of passing in a optimizer, we pass in a optimizer generator (that's what I did in my PyTorch implementation) like this:
`auto opt = [](auto& model) { return StandardSGD<decltype(model)>(model, 0.0001, 2); }`
The drawback is user has to write this lambda function and the code will become fairly overstaffed so I don't want to do this in C++.
Another delayed binding solution is to add a overload for `Optimize` of optimizer like this:
`double Optimize(arma::mat& iterate, FunctionType& function);`
It looks like the easiest way. But the problem is if we pass in the `function` every time we call the function, why do we still store it in constructor? If we do this, do we need to change the original constructor? Although this, totally separating optimizer and network, is the most efficient design I think, I don't know what's the influence of this change on existing codebase.
Rebinding is also a solution, we can refactor the optimizer with `reference_wrapper` like this:
```
template<typename FunctionType>
class GradientDescent
{
 public:
  GradientDescent(FunctionType& function,
                  const double stepSize = 0.01,
                  const size_t maxIterations = 100000,
                  const double tolerance = 1e-5);

  //! Get the instantiated function to be optimized.
  const FunctionType& Function() const { return function.get(); }
  //! Modify the instantiated function.
  FunctionType& Function() { return function.get(); }

  void Rebind(std::reference<FunctionType> f) { function = f; };

 private:
  //! The instantiated function.
  std::reference_wrapper<FunctionType> function;
};
```
This would be totally compatible with existing codebase.
If I do all the things from scratch, I'd prefer the second solution. But if refactoring the constructor of optimizer class will lead to too many regression issues, I'm fine to implement the third one.